### PR TITLE
feat: Use trim as a 3 pos toggle switch source

### LINKED ISSUE #4453: Ability to use inactive trims as 3 way switches
### Is there an existing issue for this feature request?

- [X] I have searched the existing issues

### Is your feature request related to a problem?

when a trim is disabled in FM, it cannot be used as a 'switch source'

### Describe the solution you'd like

Ability to use it. There are questions to be thought through as in essence, that switch wont be always existing, what do we do when it doesnt

### Describe alternatives you've considered

put more mixer lines instead of a simple one

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by philmoz**
Fixes #4453 

If trim is disabled in flight mode, treat it as a 3 position toggle switch in 'getValue' function.

### CODE REVIEW

**Comment by 3djc** [comment]
Could we not merge this one until I've come home ? I would like to do some tests if possible.

**Comment by 3djc** [comment]
This approach is both simple and efficient.

It does one caveat tho, if trim channel value was used somewhere in mix lines, it was returning 0 when disabled in FM, now it will return trim switch position.

Example:
you set your t5 trim channel to be the source of an += mix line on one of your aileron (in case of 2 ail per wing) to act as flaps.
When running  a FM where T5 is activated, you have flaps down and can adjust flaps value with t5
When running a FM without T5 activated, your flaps are up

But with this change, if you push T5 in a FM without flaps, they will suddenly deploy full.

I have mixed feelings, I like the simplicity of this, but I'm fearful of the possible impact on user setup

**Comment by philmoz** [comment]
> But with this change, if you push T5 in a FM without flaps, they will suddenly deploy full.

I'm not seeing this?
The original fix I posted in issue 4453 has this problem; but the version in this PR should not do that.

**Comment by 3djc** [comment]
For t20 (easy to use t5)

[model00.zip](https://github.com/EdgeTX/edgetx/files/13793544/model00.zip)

No cust switch, no flaps
Push SW6 (top left), you move to flp flight mode, you can adjust flap position with t5 (ch5)
Desactivate, SW6 (so no flaps), push T5, ch5 goes min or max

**Comment by philmoz** [comment]
> For t20 (easy to use t5)

Thanks - testing with the correct branch (when awake) helps :)

I've added a '3POS' mode to the trim settings in flight modes to explicitly change a trim switch to a 3 position toggle. This way existing setups with trim turned off will not be affected.

Screenshot 2023-12-30 at 9 23 34 am

Screenshot 2023-12-30 at 9 24 07 am

Screenshot 2023-12-30 at 9 22 54 am

**Comment by 3djc** [comment]
Very very nice, works very well now

File: radio/src/mixer.cpp [hunk 1]
```diff
@@ -384,6 +384,13 @@ getvalue_t getValue(mixsrc_t i, bool* valid)
 
   else if (i <= MIXSRC_LAST_TRIM) {
     i -= MIXSRC_FIRST_TRIM;
+    if (getRawTrimValue(mixerCurrentFlightMode, i).mode == TRIM_MODE_3POS) {
+      // Trim set as 3POS toggle switch in FM
+      uint8_t tidx = i * 2;
+      if (trimDown(tidx)) return -1024;
+      else if (trimDown(tidx + 1)) return 1024;
```
Comment (raphaelcoeffic) [inline radio/src/mixer.cpp:391]: Just nit-picking here, but didn't we have some constant for such min/max values?

Comment (philmoz) [reply radio/src/mixer.cpp]: Thanks, fixed now.

**Review by raphaelcoeffic — APPROVED** [review]
Really simple and efficient! People will love it!

**Comment by pfeerick** [comment]
You wouldn't *all* happen to be mode 1 flyers (meaning y'all have your elevator and throttle sticks the wrong way around 🤭 ) now would you? As you may want to check which trim triggers what in this case 🤔 

image
image

Any chance to teaching `companion\src\modelprinter.cpp` about the new trim mode also? 

Otherwise LGTM on TX16S, Boxer.

**Comment by philmoz** [comment]
> You wouldn't _all_ happen to be mode 1 flyers (meaning y'all have your elevator and throttle sticks the wrong way around 🤭 ) now would you? As you may want to check which trim triggers what in this case 🤔

Not sure I follow - is there a problem? 
 
> Any chance to teaching `companion\src\modelprinter.cpp` about the new trim mode also?

Done.

**Comment by 3djc** [comment]
> You wouldn't _all_ happen to be mode 1 flyers (meaning y'all have your elevator and throttle sticks the wrong way around 🤭 ) now would you? As you may want to check which trim triggers what in this case 🤔
> 
> image image
> 
> Any chance to teaching `companion\src\modelprinter.cpp` about the new trim mode also?
> 
> Otherwise LGTM on TX16S, Boxer.

Not understanding either. Not the proper way of using it in 3p mode

**Comment by pfeerick** [comment]
> Not understanding either. Not the proper way of using it in 3p mode

And does that matter? It should (and does) work there as a momentary trigger for a LS in 3P mode, which you don't get when you disable the trims. Meaning you can now also do things like sticky switch logic with the trim switches. 

> Not sure I follow - is there a problem?

Sorry, I had to leave before I could add what is probably the  pertinent bit of info... I'm mode 2. So I would like it if when setting the throttle trim to be 3P that I don't have to move the elevator trim to make this work ;)

**Comment by philmoz** [comment]
> Sorry, I had to leave before I could add what is probably the pertinent bit of info... I'm mode 2. So I would like it if when setting the throttle trim to be 3P that I don't have to move the elevator trim to make this work ;) But when used as a mix source it is checking the correct trim switch, so it seems to be specific to the logical switch.

See if the latest commit works better.

**Comment by pfeerick** [comment]
Thank you, thank you! :) Perfecto... I was hoping was just a missed mapping check! :) I've now tested modes 1-4 (since I clearly left myself wide open for that one 😆 )... logical switch tests against E and R trims were correct for each mode, as was T and A as mixer switches.

**Comment by Kevltan** [comment]
Asked for this ages ago for my turbine as regards steps. I have not tried this yet. With a throttle trim in mind and from low trim, would I be right in saying 1 click up is centre trim and 2 clicks up is full trim and the opposite applies?

**Comment by pfeerick** [comment]
No, it's a 3 position momentary switch (i.e. like the SF toggle switch). So when used as a source, it will either be high, middle or low, depending on whether you are hold the switch up (high), it is resting (middle), or holding it down (low). i.e. it reflects the physical state of the trim switch(es). For what you want, you could probably do that with a GV, using the trims to adjust it via a SF.

**Comment by Kevltan** [comment]
So its like a 3 position switch sprung to centre. So if your not pressing and holding up or down, left or right the centre position is always true?

**Comment by pfeerick** [comment]
Exactly

**Comment by Kevltan** [comment]
So why can't we have a trim option be made that can that can perform the same way I described above where you have a 2 point move - bottom, centre and top etc., so 1053 0 -1053 ? Or add an option in the coarse, medium, fine, extra  fine for axample in the drop down menu for the trims? Would this be difficult to code?

**Comment by pfeerick** [comment]
You probably don't want it on the trim steps option, unless you want the associated channel altered. But with this being added as another trim mode, adding another such as a sticky 3POS trim mode may be possible, if there is a need. Open a new issue requesting that ;)

**Comment by chakflying** [comment]
I love you! This is exactly what I needed. Unfortunately, I quickly ran out of buttons for the "USB Joystick" Advanced mode, trying to map all 4 trim switches of the T-20. I had to waste 1 virtual button for each axis for the center using the "Button - Normal - 3POS" mode. Since the center button is constantly being pressed, I also can't recycle it for something else. 

It would be great if when 2POS mode is selected, it would use <0 and >0 as the 2 button triggers, instead of mapping 0 to one of the buttons, leading to the other side being unusable.

But maybe this is a bit off-topic as an USB Joystick issue.

**Pull request merged**